### PR TITLE
Avoid duplicate generated subcomponents

### PR DESCRIPTION
**Pull request by vRallev**
This PR solves an issue that the same subcomponent is generated, what results in very surprising results. We now make sure that each generated subcomponent is unique.

### CODE REVIEW

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt [hunk 1]
```diff
@@ -295,26 +305,13 @@ internal class ContributesSubcomponentHandlerGenerator(
...
-          .addMember("%T::class", contribution.parentScope.asClassName(module))
-          .apply {
-            if (parentComponentInterface != null) {
-              addMember(
-                "replaces = [%T::class]", parentComponentInterface.componentInterface
-              )
-            }
-          }
-          .build()
-      )
```
Comment (vRallev) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGenerator.kt:310]: This no longer worked, unfortunately.

File: compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt [hunk 2]
```diff
@@ -1060,9 +1083,128 @@ class ContributesSubcomponentHandlerGeneratorTest {
     }
   }
 
-  private val Class<*>.anvilComponent: Class<*>
-    get() = classLoader
-      .loadClass("$COMPONENT_PACKAGE_PREFIX.${generatedClassesString()}$ANVIL_SUBCOMPONENT_SUFFIX")
+  @Test fun `subcomponent can be contributed and bindings replaced in a 2nd compilation`() {
+    // This test simulates a compilation in the main source set and test/androidTest source set,
+    // where contributed subcomponents are generated a second time for the test components. This
+    // test ensures that there are no duplicate generated classes.
```
Comment (vRallev) [inline compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt:1089]: Without these changes this test failed.

File: compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt [hunk 3]
```diff
@@ -1060,9 +1083,128 @@ class ContributesSubcomponentHandlerGeneratorTest {
...
+      import com.squareup.anvil.annotations.MergeComponent
+
+      @ContributesSubcomponent(Unit::class, parentScope = Any::class)
+      interface SubcomponentInterface
+
+      @MergeComponent(Any::class)
+      interface ComponentInterface1
+
+      @MergeComponent(Any::class)
+      interface ComponentInterface2
```
Comment (vRallev) [inline compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt:1184]: This test highlights the issue as well that had to be solved. Originally, we would have generated only one subcomponent, but have two parent components. That would be fine, Dagger can handle this. But this becomes problematic if there are different bindings in the parent component (that's what the test above is verifying).

File: compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt [hunk 4]
```diff
@@ -1060,9 +1083,128 @@ class ContributesSubcomponentHandlerGeneratorTest {
...
+        import com.squareup.anvil.annotations.ContributesSubcomponent
+        import com.squareup.anvil.annotations.ContributesTo
+        import com.squareup.anvil.annotations.MergeComponent
+        import dagger.Module
+        import dagger.Provides
+
+        @ContributesTo(Any::class, replaces = [DaggerModule1::class])
+        @Module
+        object DaggerModule2 {
+          @Provides fun provideIntegerSix(): Int = 6
```
Comment (vRallev) [inline compiler/src/test/java/com/squareup/anvil/compiler/codegen/ContributesSubcomponentHandlerGeneratorTest.kt:1144]: Notice that this Dagger modules replaces the other Dagger modules and the new generated subcomponent picks up the replaced module. That's quite common in tests.

**Review by 0xMatthewGroves — APPROVED** [review]
Changes make sense to me. 👍

**Pull request merged**